binding.cc

```diff
@@ -0,0 +1,1792 @@
+      n->frame_len = -1;
+    } else if (n->read_buf_free < 2 * stream->mtu) {
+      n->frame_len -= read;
+    } else {
+      return; // wait for more data
+    }
+  }
+
+  js_env_t *env = n->env;
+
+  js_handle_scope_t *scope;
+  int err = js_open_handle_scope(env, &scope);
+  assert(err == 0);
+
+  js_receiver_t ctx;
+  err = js_get_reference_value(env, n->ctx, ctx);
+  assert(err == 0);
+
+  cb_stream_data_t callback;
+  err = js_get_reference_value(env, n->on_data, callback);
+  assert(err == 0);
+
+  js_value_t *argv[1];
+  err = js_create_uint32(env, read, &(argv[0]));
+  assert(err == 0);
+
+  js_typedarray_span_t<> res;
+  err = js_call_function_with_checkpoint(env, callback, ctx, static_cast<uint32_t>(read), res);
+
+  if (err == 0) {
+    n->read_buf = res.data();
+    n->read_buf_free = res.size();
+    n->read_buf_head = n->read_buf;
+  } else {
+    // avoid re-entry
+    if (!(n->udx->exiting)) {
+      js_handle_scope_t *inner_scope;
+      err = js_open_handle_scope(env, &inner_scope);
+      assert(err == 0);
+
+      js_receiver_t ctx;
+      err = js_get_reference_value(env, n->ctx, ctx);
+      assert(err == 0);
+
+      cb_stream_realloc_data_t callback;
+      err = js_get_reference_value(env, n->realloc_data, callback);
+      assert(err == 0);
+
+      err = js_call_function_with_checkpoint(env, callback, ctx, res);
+      assert(err == 0);
+
+      n->read_buf = res.data();
+      n->read_buf_free = res.size();
+      n->read_buf_head = n->read_buf;
+
+      err = js_close_handle_scope(env, inner_scope);
+      assert(err == 0);
+    }
+  }
+
+  err = js_close_handle_scope(env, scope);
+  assert(err == 0);
+}
+
+static void
+on_udx_stream_drain (udx_stream_t *stream) {
+  auto n = reinterpret_cast<udx_napi_stream_t *>(stream);
+  if (n->udx->exiting) return;
+
+  js_env_t *env = n->env;
+
+  js_handle_scope_t *scope;
+  int err = js_open_handle_scope(env, &scope);
+  assert(err == 0);
+
+  js_receiver_t ctx;
+  err = js_get_reference_value(env, n->ctx, ctx);
+  assert(err == 0);
+
+  cb_stream_drain_t callback;
+  err = js_get_reference_value(env, n->on_drain, callback);
+  assert(err == 0);
+
+  err = js_call_function_with_checkpoint(env, callback, ctx);
+  assert(err != js_pending_exception);
+
+  err = js_close_handle_scope(env, scope);
+  assert(err == 0);
+}
+
+static void
+on_udx_stream_ack (udx_stream_write_t *req, int status, int unordered) {
+  int err = 0;
+
+  auto n = reinterpret_cast<udx_napi_stream_t *>(req->stream);
+  if (n->udx->exiting) return;
+
+  js_env_t *env = n->env;
+
+  js_handle_scope_t *scope;
+  err = js_open_handle_scope(env, &scope);
+  assert(err == 0);
+
+  js_receiver_t ctx;
+  err = js_get_reference_value(env, n->ctx, ctx);
+  assert(err == 0);
+
+  cb_stream_ack_t callback;
+  err = js_get_reference_value(env, n->on_ack, callback);
+  assert(err == 0);
+
+  auto offset = (uintptr_t) req->data; // TODO: search uintptr_t and consider replacing with native pointers
```

Comment:
```suggestion
  auto offset = reinterpret_cast<uintptr_t>(req->data); // TODO: search uintptr_t and consider replacing with native pointers
```

Reply:
There are some other C-style casts as well that should be changed to appropriate C++ casts.